lib/eventHubReceiver.ts

```diff
@@ -0,0 +1,382 @@
+        }
+      }
+    });
+
+    this.on("removeListener", (event) => {
+      if (event === Constants.message) {
+        if (this._session && this._receiver) {
+          debug("Removing an event handler for the 'message' event on the underlying amqp receiver: ", this.name!);
+          this._receiver.removeListener(Constants.message, onMessage);
+        }
+      }
+
+      if (event === Constants.error) {
+        if (this._session && this._receiver) {
+          debug("Removing an event handler for the 'receiver_error' event on the underlying amqp receiver: ", this.name!);
+          this._receiver.removeListener(Constants.receiverError, onError);
+        }
+      }
+    });
+  }
+
+  /**
+   * Creates a new AMQP receiver under a new AMQP session.
+   * @returns {Promoise<void>}
+   */
+  async init(): Promise<void> {
+    try {
+      // Acquire the lock and establish a cbs session if it does not exist on the connection.
+      await defaultLock.acquire(this._context.cbsSession.cbsLock, () => { return this._context.cbsSession.init(this._context.connection); });
+      const tokenObject = await this._context.tokenProvider.getToken(this.audience);
+      debug(`[${this._context.connectionId}] EH Receiver: calling negotiateClaim for audience "${this.audience}"`);
+      // Negotitate the CBS claim.
+      await this._context.cbsSession.negotiateClaim(this.audience, this._context.connection, tokenObject);
+      if (!this._session && !this._receiver) {
+        let receiverError: any;
+        let rcvrOptions: rheaPromise.ReceiverOptions = {
+          autoaccept: true,
+          source: {
+            address: this.address
+          },
+          credit_window: this.prefetchCount,
+        };
+        if (this.epoch !== undefined && this.epoch !== null) {
+          if (!rcvrOptions.properties) rcvrOptions.properties = {};
+          rcvrOptions.properties[Constants.attachEpoch] = rhea.types.wrap_long(this.epoch);
+        }
+        if (this.identifier) {
+          if (!rcvrOptions.properties) rcvrOptions.properties = {};
+          rcvrOptions.properties[Constants.receiverIdentifierName] = this.identifier;
+        }
+        if (this.receiverRuntimeMetricEnabled) {
+          rcvrOptions.desired_capabilities = Constants.enableReceiverRuntimeMetricName;
+        }
+        if (this.options && this.options.eventPosition) {
+          // Set filter on the receiver if event position is specified.
+          let filterClause = this.options.eventPosition.getExpression();
+          if (filterClause) {
+            rcvrOptions.source.filter = {
+              "apache.org:selector-filter:string": rhea.types.wrap_described(filterClause, 0x468C00000004)
+            };
+          }
+        }
+        this._session = await rheaPromise.createSession(this._context.connection);
+        const handleReceiverError = (context: rheaPromise.Context) => {
+          receiverError = errors.translate(context.receiver.error);
```

Comment:
I see. Will keep that in mind and add a TODO in the source code. The problem is that at times the error can happen early enough where I am not able to cache the name of the receiver (since this happens during receiver creation) there is no definite way for me to know if it was the same receiver).

I had to add this error handler at the session level because at times it does so happen that the service sends an attach frame and the promise of creating the receiver is resolved since rhea emits receiver_open event and moments later it sends a detach frame. While it sends the detach frame and this promise inside the EventHubReceiver is not yet resolved then I catch that error and reject this promise.